Add type checking to CI

sergei-maertens opened the pull request:
This setup with tox.ini *should* make the mypy checks run after the (unit) tests for each python/django env.

sergei-maertens committed:
:construction_worker: Add mypy checking to tox

sergei-maertens committed:
x509 name values can be string or bytes, not just string

sergei-maertens committed:
:label: Type hints in model mixins

Type annotating mixins is the worst. Using a Protocol doesn't
work because the  expects a model
class to make the _meta work, and just overriding the
self type also doesn't work properly.

Viicos commented on simple_certmanager/utils.py:
```diff
@@ -19,8 +19,20 @@ def load_pem_x509_private_key(data: bytes):
     return load_pem_private_key(data, password=None)
 
 
+def _decode(value: Union[str, bytes]) -> str:
+    # attr.value can be bytes, in which case it is must be an UTF8String or
+    # PrintableString (the latter being a subset of ASCII, thus also a subset of UTF8)
+    # See https://www.rfc-editor.org/rfc/rfc5280.txt
+    if not isinstance(value, bytes):
+        # https://github.com/microsoft/pyright/issues/6059 Pyright does not like this :(
+        return value
+    return value.decode("utf8")
```
Why not `django.utils.text.smart/force_str`?

sergei-maertens replied:
Yeah I wasn't sure if it was appropriate because those utils aren't limited to `str` or `bytes`, but we should only get those types so I've updated it.

sergei-maertens committed:
:label: Make type checking pass with mypy

* the admin.register decorator does not provide the generic argument,
  so we need to specify the model type explicitly
* testutils needs to be added to type check dependencies so the import
  can be resolved
* ignore type checking on factory boy until they ship type hints in
  the package, which is being worked on

sergei-maertens commented on pyrightconfig.json:
```diff
@@ -0,0 +1,3 @@
+{
+    "disableBytesTypePromotions": true
```
without this setting, pyright turns `bytes` into `bytes | bytearray | memoryview`, which blows up our code because only `bytes` has a `decode` method.

The return type of cryptography for x509 name values is `str | bytes`, never was there an intent to support bytearray or memoryview

sergei-maertens commented on simple_certmanager/admin.py:
```diff
@@ -10,6 +10,7 @@
 
 @admin.register(Certificate)
 class CertificateAdmin(PrivateMediaMixin, admin.ModelAdmin):
+    model: type[Certificate]
```
even though the model is defined as generic in both the mixin and the stubs for `ModelAdmin`, this is not picked up properly - I suspect because of the `@admin.register` usage which uses different type vars.

Declaring this type explicitly makes mypy accept it :shrug:

Viicos replied:
Checked and the `register` decorator works with both mypy and pyright. It is not picked up properly because `PrivateMediaMixin` defines `model: type[Model]`, overriding the definition from `ModelAdmin` as it comes first in the mro.

And even then, to make it work properly you would have to inherit form `admin.ModelAdmin[Certificate]`, assuming you make it subscriptable (see https://github.com/lazybird/django-solo/pull/128#discussion_r1451790020). So it's a pain :/

sergei-maertens replied:
I've actually just patched django-privates to do the same thing as django-stubs with a type var: https://github.com/sergei-maertens/django-privates/blob/27c7391ef82479f5d99fb6542647ab999b3ebd2a/privates/admin.py#L46 so it no longer defines `model: type[Model]`

Also made the mixin generic and tried `PrivateMediaMixin[Certificate]`, which didn't yield any results. `admin.ModelAdmin` is not subscriptable out of the box, so the shit you have to do to make this pass is a bit excessive...

sergei-maertens committed:
:label: Improve type hints for decorator

sergei-maertens committed:
:ok_hand: Use django.utils.encoding.force_str

sergei-maertens committed:
:rotating_light: flake8

Viicos committed:
:label: couple type improvements

Viicos reviewed (approved):
Still far from mypy strict but this looks nice
